From 0ff79ec866c3a86a29fa37bd0745b70bd886a667 Mon Sep 17 00:00:00 2001
From: Mathieu Audat <mathieu.audat@savoirfairelinux.com>
Date: Thu, 16 Nov 2017 10:50:27 +0200
Subject: [PATCH 46/47] remove Mesh link in wpa_supplicant_event_assoc only if
 CONFIG_MESH.

Signed-off-by: Mathieu Audat <mathieu.audat@savoirfairelinux.com>
---
 wpa_supplicant/events.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/wpa_supplicant/events.c b/wpa_supplicant/events.c
index 4c3171e..4c843b0 100755
--- a/wpa_supplicant/events.c
+++ b/wpa_supplicant/events.c
@@ -3040,6 +3040,7 @@ static void wpa_supplicant_event_assoc(struct wpa_supplicant *wpa_s,
 	/*
 	* after we finished handling the STA connection we can now remove the Mesh link
 	*/
+#ifdef CONFIG_MESH
 	if ( (wpa_s->global->mesh_on_demand.enabled)
 		&& (wpa_s->wpa_state == WPA_COMPLETED)
 		&&  (wpa_s->global->mesh_on_demand.anyMeshConnected)) {
@@ -3047,6 +3048,7 @@ static void wpa_supplicant_event_assoc(struct wpa_supplicant *wpa_s,
 			mesh_mpm_close_links(wpa_s->global->mesh_on_demand.mesh_wpa_s,
 								     wpa_s->global->mesh_on_demand.mesh_wpa_s->ifmsh);
 	}
+#endif /* CONFIG_MESH */
 }
 
 
-- 
1.9.1

